[GTK] Move gtk specific code from PlatformDisplay to a new class

carlosgcampos opened the pull request:
#### 8db1e051cf4a6d7877dd51b81b478e8367d3f052
<pre>
[GTK] Move gtk specific code from PlatformDisplay to a new class
<a href="https://bugs.webkit.org/show_bug.cgi?id=278273">https://bugs.webkit.org/show_bug.cgi?id=278273</a>

Reviewed by Michael Catanzaro.

There&apos;s gtk specific code in PlatformDisplay that is only used in the UI
process. We can move that to a new Display class in WebKit layer and
remove also PlatformDisplayX11 and PlatformDisplayWayland, since they
are now unused in the web process. The remaining PlatformDisplay
implementations are only used for accelerated compositing, and they
always initialize EGL in the constructor, so we can simplify the EGL
initialization and make it fatal if there&apos;s a failure.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::sharedDisplay):
(WebCore::PlatformDisplay::PlatformDisplay):
(WebCore::PlatformDisplay::~PlatformDisplay):
(WebCore::PlatformDisplay::eglDisplay const):
(WebCore::PlatformDisplay::eglCheckVersion const):
(WebCore::PlatformDisplay::eglExtensions const):
(WebCore::PlatformDisplay::terminateEGLDisplay):
(WebCore::PlatformDisplay::createEGLImage const):
(WebCore::PlatformDisplay::destroyEGLImage const):
(WebCore::PlatformDisplay::dmabufFormats):
(WebCore::PlatformDisplay::createPlatformDisplay): Deleted.
(WebCore::PlatformDisplay::sharedDisplayDidClose): Deleted.
(WebCore::PlatformDisplay::initializeEGLDisplay): Deleted.
(WebCore::PlatformDisplay::accessibilityBusAddress const): Deleted.
* Source/WebCore/platform/graphics/PlatformDisplay.h:
(WebCore::PlatformDisplay::gtkEGLDisplay): Deleted.
* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::createWindowContext):
(WebCore::GLContext::createOffscreen):
(WebCore::GLContext::createSharing):
* Source/WebCore/platform/graphics/egl/PlatformDisplaySurfaceless.cpp:
(WebCore::PlatformDisplaySurfaceless::create):
(WebCore::PlatformDisplaySurfaceless::PlatformDisplaySurfaceless):
* Source/WebCore/platform/graphics/egl/PlatformDisplaySurfaceless.h:
* Source/WebCore/platform/graphics/gbm/PlatformDisplayGBM.cpp:
(WebCore::PlatformDisplayGBM::create):
(WebCore::PlatformDisplayGBM::PlatformDisplayGBM):
* Source/WebCore/platform/graphics/gbm/PlatformDisplayGBM.h:
* Source/WebCore/platform/graphics/libwpe/PlatformDisplayLibWPE.cpp:
(WebCore::PlatformDisplayLibWPE::create):
(WebCore::PlatformDisplayLibWPE::PlatformDisplayLibWPE):
(WebCore::PlatformDisplayLibWPE::initializeEGLDisplay): Deleted.
* Source/WebCore/platform/graphics/libwpe/PlatformDisplayLibWPE.h:
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp: Removed.
* Source/WebCore/platform/graphics/win/PlatformDisplayWin.cpp:
(WebCore::PlatformDisplayWin::create):
(WebCore::PlatformDisplayWin::PlatformDisplayWin):
(WebCore::PlatformDisplayWin::initializeEGLDisplay): Deleted.
* Source/WebCore/platform/graphics/win/PlatformDisplayWin.h:
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp: Removed.
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.h: Removed.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/InputMethodFilter.cpp:
(WebKit::InputMethodFilter::isViewFocused const):
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::uiProcessContextIsEGL):
(WebKit::WebKitProtocolHandler::handleGPU):
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
(UIClient::windowConfigureEventCallback):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseSetToplevelOnScreenWindow):
(webkitWebViewBaseIsFocused):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewGtk.cpp:
(webkitWebViewMaximizeWindow):
(webkitWebViewRestoreWindow):
* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bindWayland):
(WebKit::shouldUnshareNetwork):
(WebKit::bubblewrapSpawn):
* Source/WebKit/UIProcess/glib/DRMDevice.cpp:
(WebKit::currentEGLDisplay):
(WebKit::drmPrimaryDevice):
(WebKit::drmRenderNodeDevice):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::accessibilityBusAddress const):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::rendererBufferMode):
(WebKit::AcceleratedBackingStoreDMABuf::preferredBufferFormats):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::create):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::~BufferEGLImage):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBuffer):
* Source/WebKit/UIProcess/gtk/Display.cpp: Added.
(WebKit::Display::singleton):
(WebKit::Display::Display):
(WebKit::Display::~Display):
(WebKit::Display::glDisplay const):
(WebKit::Display::accessibilityBusAddress const):
* Source/WebKit/UIProcess/gtk/Display.h: Added.
(WebKit::Display::glDisplayIsSharedWithGtk const):
* Source/WebKit/UIProcess/gtk/DisplayWayland.cpp: Added.
(WebKit::Display::isWayland const):
(WebKit::Display::initializeGLDisplayWayland const):
* Source/WebKit/UIProcess/gtk/DisplayX11.cpp: Added.
(WebKit::Display::isX11 const):
(WebKit::Display::initializeGLDisplayX11 const):
(WebKit::Display::accessibilityBusAddressX11 const):
* Source/WebKit/UIProcess/gtk/PointerLockManager.cpp:
(WebKit::PointerLockManager::create):

Canonical link: <a href="https://commits.webkit.org/282416@main">https://commits.webkit.org/282416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97acd39c8ce9520a3f0b51c71fccf6b592a9fef6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63085 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67106 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13689 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65205 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50128 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13973 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50836 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9447 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66154 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54626 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31522 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36104 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11973 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12565 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57644 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12302 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68801 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7031 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11914 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58149 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58401 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5865 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9518 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38261 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39341 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/PlatformDisplay.h:
```diff
@@ -124,26 +108,10 @@ class PlatformDisplay {
-
-    GRefPtr<GdkDisplay> m_sharedDisplay;
-#endif
+    PlatformDisplay(std::unique_ptr<GLDisplay>&&);
```
```suggestion
    explicit PlatformDisplay(std::unique_ptr<GLDisplay>&&);
```

mcatanzaro commented on Source/WebCore/platform/graphics/egl/PlatformDisplaySurfaceless.cpp:
```diff
@@ -33,19 +33,26 @@ namespace WebCore {
 
-    PlatformDisplay::initializeEGLDisplay();
+    if (!glDisplay) {
+        WTFLogAlways("Could not create EGL display: %s. Aborting...\n", GLContext::lastErrorString());
```
```suggestion
        WTFLogAlways("Could not create EGL display: %s. Aborting...", GLContext::lastErrorString());
```

mcatanzaro commented on Source/WebCore/platform/graphics/gbm/PlatformDisplayGBM.cpp:
```diff
@@ -35,19 +35,24 @@ namespace WebCore {
 
-    PlatformDisplay::initializeEGLDisplay();
+    if (!glDisplay) {
+        WTFLogAlways("Could not create EGL display: %s. Aborting...\n", GLContext::lastErrorString());
```
```suggestion
        WTFLogAlways("Could not create EGL display: %s. Aborting...", GLContext::lastErrorString());
```

mcatanzaro commented on Source/WebCore/platform/graphics/libwpe/PlatformDisplayLibWPE.cpp:
```diff
@@ -87,24 +81,30 @@ PlatformDisplayLibWPE::PlatformDisplayLibWPE(int hostFd)
+        glDisplay = GLDisplay::create(eglGetDisplay(eglNativeDisplay));
+
+    if (!glDisplay) {
+        WTFLogAlways("Could not create EGL display: %s. Aborting...\n", GLContext::lastErrorString());
```
```suggestion
        WTFLogAlways("Could not create EGL display: %s. Aborting...", GLContext::lastErrorString());
```

mcatanzaro commented on Source/WebCore/platform/graphics/win/PlatformDisplayWin.cpp:
```diff
@@ -41,9 +42,18 @@ void PlatformDisplayWin::initializeEGLDisplay()
-    m_eglDisplay = GLDisplay::create(eglGetPlatformDisplayEXT(EGL_PLATFORM_ANGLE_ANGLE, EGL_DEFAULT_DISPLAY, attributes));
+    auto glDisplay = GLDisplay::create(eglGetPlatformDisplayEXT(EGL_PLATFORM_ANGLE_ANGLE, EGL_DEFAULT_DISPLAY, attributes));
+    if (!glDisplay) {
+        WTFLogAlways("Could not create EGL display: %s. Aborting...\n", GLContext::lastErrorString());
```
```suggestion
        WTFLogAlways("Could not create EGL display: %s. Aborting...", GLContext::lastErrorString());
```

mcatanzaro commented on Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
```diff
@@ -254,6 +257,8 @@ static void bindWayland(Vector<CString>& args)
     GUniquePtr<char> waylandRuntimeFile(g_build_filename(runtimeDir, display, nullptr));
     bindIfExists(args, waylandRuntimeFile.get(), BindFlags::ReadWrite);
+
+    args.append("--unshare-ipc");
```
Nit: this was better where it used to be, since it's something that's done when using Wayland rather than something that's done as part of binding the Wayland socket.

carlosgcampos committed:
[GTK] Move gtk specific code from PlatformDisplay to a new class
https://bugs.webkit.org/show_bug.cgi?id=278273

Reviewed by Michael Catanzaro.

There's gtk specific code in PlatformDisplay that is only used in the UI
process. We can move that to a new Display class in WebKit layer and
remove also PlatformDisplayX11 and PlatformDisplayWayland, since they
are now unused in the web process. The remaining PlatformDisplay
implementations are only used for accelerated compositing, and they
always initialize EGL in the constructor, so we can simplify the EGL
initialization and make it fatal if there's a failure.

* Source/WebCore/PlatformGTK.cmake:
* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::sharedDisplay):
(WebCore::PlatformDisplay::PlatformDisplay):
(WebCore::PlatformDisplay::~PlatformDisplay):
(WebCore::PlatformDisplay::eglDisplay const):
(WebCore::PlatformDisplay::eglCheckVersion const):
(WebCore::PlatformDisplay::eglExtensions const):
(WebCore::PlatformDisplay::terminateEGLDisplay):
(WebCore::PlatformDisplay::createEGLImage const):
(WebCore::PlatformDisplay::destroyEGLImage const):
(WebCore::PlatformDisplay::dmabufFormats):
(WebCore::PlatformDisplay::createPlatformDisplay): Deleted.
(WebCore::PlatformDisplay::sharedDisplayDidClose): Deleted.
(WebCore::PlatformDisplay::initializeEGLDisplay): Deleted.
(WebCore::PlatformDisplay::accessibilityBusAddress const): Deleted.
* Source/WebCore/platform/graphics/PlatformDisplay.h:
(WebCore::PlatformDisplay::gtkEGLDisplay): Deleted.
* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::createWindowContext):
(WebCore::GLContext::createOffscreen):
(WebCore::GLContext::createSharing):
* Source/WebCore/platform/graphics/egl/PlatformDisplaySurfaceless.cpp:
(WebCore::PlatformDisplaySurfaceless::create):
(WebCore::PlatformDisplaySurfaceless::PlatformDisplaySurfaceless):
* Source/WebCore/platform/graphics/egl/PlatformDisplaySurfaceless.h:
* Source/WebCore/platform/graphics/gbm/PlatformDisplayGBM.cpp:
(WebCore::PlatformDisplayGBM::create):
(WebCore::PlatformDisplayGBM::PlatformDisplayGBM):
* Source/WebCore/platform/graphics/gbm/PlatformDisplayGBM.h:
* Source/WebCore/platform/graphics/libwpe/PlatformDisplayLibWPE.cpp:
(WebCore::PlatformDisplayLibWPE::create):
(WebCore::PlatformDisplayLibWPE::PlatformDisplayLibWPE):
(WebCore::PlatformDisplayLibWPE::initializeEGLDisplay): Deleted.
* Source/WebCore/platform/graphics/libwpe/PlatformDisplayLibWPE.h:
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp: Removed.
* Source/WebCore/platform/graphics/win/PlatformDisplayWin.cpp:
(WebCore::PlatformDisplayWin::create):
(WebCore::PlatformDisplayWin::PlatformDisplayWin):
(WebCore::PlatformDisplayWin::initializeEGLDisplay): Deleted.
* Source/WebCore/platform/graphics/win/PlatformDisplayWin.h:
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.cpp: Removed.
* Source/WebCore/platform/graphics/x11/PlatformDisplayX11.h: Removed.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/API/glib/InputMethodFilter.cpp:
(WebKit::InputMethodFilter::isViewFocused const):
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::uiProcessContextIsEGL):
(WebKit::WebKitProtocolHandler::handleGPU):
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
(UIClient::windowConfigureEventCallback):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseSetToplevelOnScreenWindow):
(webkitWebViewBaseIsFocused):
* Source/WebKit/UIProcess/API/gtk/WebKitWebViewGtk.cpp:
(webkitWebViewMaximizeWindow):
(webkitWebViewRestoreWindow):
* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bindWayland):
(WebKit::shouldUnshareNetwork):
(WebKit::bubblewrapSpawn):
* Source/WebKit/UIProcess/glib/DRMDevice.cpp:
(WebKit::currentEGLDisplay):
(WebKit::drmPrimaryDevice):
(WebKit::drmRenderNodeDevice):
* Source/WebKit/UIProcess/glib/WebProcessPoolGLib.cpp:
(WebKit::WebProcessPool::accessibilityBusAddress const):
* Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp:
(WebKit::AcceleratedBackingStoreDMABuf::rendererBufferMode):
(WebKit::AcceleratedBackingStoreDMABuf::preferredBufferFormats):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::create):
(WebKit::AcceleratedBackingStoreDMABuf::BufferEGLImage::~BufferEGLImage):
(WebKit::AcceleratedBackingStoreDMABuf::didCreateBuffer):
* Source/WebKit/UIProcess/gtk/Display.cpp: Added.
(WebKit::Display::singleton):
(WebKit::Display::Display):
(WebKit::Display::~Display):
(WebKit::Display::glDisplay const):
(WebKit::Display::accessibilityBusAddress const):
* Source/WebKit/UIProcess/gtk/Display.h: Added.
(WebKit::Display::glDisplayIsSharedWithGtk const):
* Source/WebKit/UIProcess/gtk/DisplayWayland.cpp: Added.
(WebKit::Display::isWayland const):
(WebKit::Display::initializeGLDisplayWayland const):
* Source/WebKit/UIProcess/gtk/DisplayX11.cpp: Added.
(WebKit::Display::isX11 const):
(WebKit::Display::initializeGLDisplayX11 const):
(WebKit::Display::accessibilityBusAddressX11 const):
* Source/WebKit/UIProcess/gtk/PointerLockManager.cpp:
(WebKit::PointerLockManager::create):

Canonical link: https://commits.webkit.org/282416@main
